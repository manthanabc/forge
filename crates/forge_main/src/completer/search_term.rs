use reedline::Span;

pub struct SearchTerm {
    line: String,
    position: usize,
}

impl SearchTerm {
    pub fn new(line: &str, position: usize) -> Self {
        if position > line.len() {
            panic!(
                "Position {position} is out of bounds: string '{line}' (length: {})",
                line.len()
            );
        }
        Self { line: line.to_string(), position }
    }

    /// Get the search term from the line based on '@' marker or cursor position
    ///
    /// If '@' marker is present, returns the word following it.
    /// Otherwise, returns the word at the cursor position.
    /// If no word is found, returns None.
    pub fn process(&self) -> Option<TermResult<'_>> {
        // Ensure position is on a UTF-8 character boundary to prevent panics
        let safe_position = if self.line.is_char_boundary(self.position) {
            self.position
        } else {
            // Find the nearest lower character boundary
            (0..self.position)
                .rev()
                .find(|&i| self.line.is_char_boundary(i))
                .unwrap_or(0)
        };

        let prefix = &self.line[..safe_position];
        let at_pos = prefix.rfind('@')?;
        let start_pos = at_pos + 1;
        let term = &self.line[start_pos..safe_position];

        Some(TermResult { span: Span::new(start_pos, safe_position), term })
    }
}

#[derive(Debug)]
pub struct TermResult<'a> {
    pub span: Span,
    pub term: &'a str,
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::SearchTerm;

    impl SearchTerm {
        fn test(line: &str) -> Vec<TermSpec> {
            // Test at each valid character boundary position, starting from 1
            (1..=line.len())
                .filter(|&pos| line.is_char_boundary(pos))
                .map(|pos| {
                    let input = SearchTerm::new(line, pos);
                    let output = input.process();
                    let (a, b) = line.split_at(pos);

                    TermSpec {
                        pos,
                        input: format!("{a}[{b}"),
                        output: output.as_ref().map(|term| term.term.to_string()),
                        span_start: output.as_ref().map(|term| term.span.start),
                        span_end: output.as_ref().map(|term| term.span.end),
                    }
                })
                .collect()
        }
    }

    #[derive(Debug)]
    #[allow(dead_code)] // Used to generate test snapshots
    struct TermSpec {
        input: String,
        output: Option<String>,
        span_start: Option<usize>,
        span_end: Option<usize>,
        pos: usize,
    }

    #[test]
    fn test_marker_based_search() {
        let results = SearchTerm::test("@abc @def ghi@");
        assert_debug_snapshot!(results);
    }

    #[test]
    fn test_marker_based_search_chinese() {
        let results = SearchTerm::test("@ä½ å¥½ @ä¸–ç•Œ æµ‹è¯•@");
        assert_debug_snapshot!(results);
    }

    #[test]
    fn test_marker_based_search_mixed_chinese_english() {
        let results = SearchTerm::test("@hello @ä¸–ç•Œ test@ä¸­æ–‡");
        assert_debug_snapshot!(results);
    }

    #[test]
    fn test_marker_based_search_chinese_with_spaces() {
        let results = SearchTerm::test("@ä¸­ æ–‡ @æµ‹è¯•");
        assert_debug_snapshot!(results);
    }

    #[test]
    fn test_marker_based_search_emoji() {
        let results = SearchTerm::test("@ðŸš€ @ðŸŒŸ emoji@");
        assert_debug_snapshot!(results);
    }
}
