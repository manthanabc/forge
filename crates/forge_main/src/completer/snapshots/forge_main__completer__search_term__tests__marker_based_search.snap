---
source: crates/forge_main/src/completer/search_term.rs
expression: results
---
[
    TermSpec {
        input: "@[abc @def ghi@",
        output: Some(
            "",
        ),
        span_start: Some(
            1,
        ),
        span_end: Some(
            1,
        ),
        pos: 1,
    },
    TermSpec {
        input: "@a[bc @def ghi@",
        output: Some(
            "a",
        ),
        span_start: Some(
            1,
        ),
        span_end: Some(
            2,
        ),
        pos: 2,
    },
    TermSpec {
        input: "@ab[c @def ghi@",
        output: Some(
            "ab",
        ),
        span_start: Some(
            1,
        ),
        span_end: Some(
            3,
        ),
        pos: 3,
    },
    TermSpec {
        input: "@abc[ @def ghi@",
        output: Some(
            "abc",
        ),
        span_start: Some(
            1,
        ),
        span_end: Some(
            4,
        ),
        pos: 4,
    },
    TermSpec {
        input: "@abc [@def ghi@",
        output: Some(
            "abc ",
        ),
        span_start: Some(
            1,
        ),
        span_end: Some(
            5,
        ),
        pos: 5,
    },
    TermSpec {
        input: "@abc @[def ghi@",
        output: Some(
            "",
        ),
        span_start: Some(
            6,
        ),
        span_end: Some(
            6,
        ),
        pos: 6,
    },
    TermSpec {
        input: "@abc @d[ef ghi@",
        output: Some(
            "d",
        ),
        span_start: Some(
            6,
        ),
        span_end: Some(
            7,
        ),
        pos: 7,
    },
    TermSpec {
        input: "@abc @de[f ghi@",
        output: Some(
            "de",
        ),
        span_start: Some(
            6,
        ),
        span_end: Some(
            8,
        ),
        pos: 8,
    },
    TermSpec {
        input: "@abc @def[ ghi@",
        output: Some(
            "def",
        ),
        span_start: Some(
            6,
        ),
        span_end: Some(
            9,
        ),
        pos: 9,
    },
    TermSpec {
        input: "@abc @def [ghi@",
        output: Some(
            "def ",
        ),
        span_start: Some(
            6,
        ),
        span_end: Some(
            10,
        ),
        pos: 10,
    },
    TermSpec {
        input: "@abc @def g[hi@",
        output: Some(
            "def g",
        ),
        span_start: Some(
            6,
        ),
        span_end: Some(
            11,
        ),
        pos: 11,
    },
    TermSpec {
        input: "@abc @def gh[i@",
        output: Some(
            "def gh",
        ),
        span_start: Some(
            6,
        ),
        span_end: Some(
            12,
        ),
        pos: 12,
    },
    TermSpec {
        input: "@abc @def ghi[@",
        output: Some(
            "def ghi",
        ),
        span_start: Some(
            6,
        ),
        span_end: Some(
            13,
        ),
        pos: 13,
    },
    TermSpec {
        input: "@abc @def ghi@[",
        output: Some(
            "",
        ),
        span_start: Some(
            14,
        ),
        span_end: Some(
            14,
        ),
        pos: 14,
    },
]
